// @shikaku/engine/src/index.ts
export type Cell = { x: number; y: number };
export type Clue = Cell & { value: number };
export type Rectangle = { x: number; y: number; width: number; height: number };

export type Board = { width: number; height: number; clues: Clue[] };

export type BoardState = {
  board: Board;
  rectangles: Rectangle[];
};

export type ValidationResult = {
  isSolved: boolean;
  errors: string[];
};

export function getCellsForRectangle(rectangle: Rectangle): Cell[] {
  const cells: Cell[] = [];
  for (let y = rectangle.y; y < rectangle.y + rectangle.height; y += 1) {
    for (let x = rectangle.x; x < rectangle.x + rectangle.width; x += 1) {
      cells.push({ x, y });
    }
  }
  return cells;
}

export function isRectangleInsideBoard(board: Board, rectangle: Rectangle): boolean {
  return rectangle.x >= 0 &&
    rectangle.y >= 0 &&
    rectangle.x + rectangle.width <= board.width &&
    rectangle.y + rectangle.height <= board.height;
}

export function areCellsEqual(c1: Cell, c2: Cell): boolean {
  return c1.x === c2.x && c1.y === c2.y;
}

export function getRectangleArea(rectangle: Rectangle): number {
  return rectangle.width * rectangle.height;
}

export function doRectanglesOverlap(r1: Rectangle, r2: Rectangle): boolean {
  return !(r1.x + r1.width <= r2.x ||
    r2.x + r2.width <= r1.x ||
    r1.y + r1.height <= r2.y ||
    r2.y + r2.height <= r1.y);
}

export function containsCell(rectangle: Rectangle, cell: Cell): boolean {
  return cell.x >= rectangle.x &&
    cell.x < rectangle.x + rectangle.width &&
    cell.y >= rectangle.y &&
    cell.y < rectangle.y + rectangle.height;
}

export function validateSolvedBoard(board: Board, rectangles: Rectangle[]): ValidationResult {
  const errors: string[] = [];
  const occupied = new Set<string>();

  rectangles.forEach((rectangle, index) => {
    // 1. Validate dimensions
    if (rectangle.width <= 0 || rectangle.height <= 0) {
      errors.push(`Rectangle ${index} has invalid dimensions: ${rectangle.width}x${rectangle.height}`);
      return; // Skip further checks for invalid shapes
    }

    // 2. Bound checks
    if (!isRectangleInsideBoard(board, rectangle)) {
      errors.push(`Rectangle ${index} is out of bounds`);
    }

    // 3. Mark cells and overlap checks
    // We check overlaps using pairs with `doRectanglesOverlap` instead of a grid matrix
    // but the set is kept to verify all board cells are covered without O(N) queries
    const cells = getCellsForRectangle(rectangle);
    for (const cell of cells) {
      const key = `${cell.x},${cell.y}`;
      occupied.add(key);
    }

    // Check overlaps explicitly using the logic helper
    for (let j = index + 1; j < rectangles.length; j++) {
      if (doRectanglesOverlap(rectangle, rectangles[j])) {
        errors.push(`Overlap detected between Rectangle ${index} and Rectangle ${j}`);
      }
    }

    const containedClues = board.clues.filter(clue => containsCell(rectangle, clue));

    if (containedClues.length !== 1) {
      errors.push(`Rectangle ${index} must contain exactly one clue`);
      return;
    }

    const area = getRectangleArea(rectangle);
    if (containedClues[0].value !== area) {
      errors.push(`Rectangle ${index} area ${area} does not match clue ${containedClues[0].value}`);
    }
  });

  for (let y = 0; y < board.height; y += 1) {
    for (let x = 0; x < board.width; x += 1) {
      const key = `${x},${y}`;
      if (!occupied.has(key)) {
        errors.push(`Uncovered cell at ${key}`);
      }
    }
  }

  return {
    isSolved: errors.length === 0,
    errors
  };
}
